### ISSUE TITLE
understanding weighted inputs functionality

### ISSUE THREAD

**Issue by mohammedayub44** [OpenNMT/OpenNMT-tf#622]
Hi,

I'm trying to use the `weighted inputs` functionality that was released recently, in the hope to see if it reduces over-fitting in domain adaption.  Documentation states we can give multiple training files which I have (presently two for now). My question is on the eval files, vocabulary and attention files (previously I have used train_alignments).
Can I just give two eval files one for each train file ? 
Do I have to combine the vocabulary sets from both train files for `source_words_vocabulary` and `target_words_vocabulary` ?
For `train_alignments` do I have to generate one or two files ?

Config file for data portion looks something like this - 

model_dir: Weighted_Attention_Run/model/

data:
  train_features_file: 
    - euro_train.en
    - custom_train.en
  train_labels_file: 
    - euro_train.es
    - custom_train.es
  eval_features_file: 
    - euro_dev.en
    - custom_dev.en
  eval_labels_file: 
    - euro_dev.es
    - custom_dev.es
  source_words_vocabulary: combined_src_vocab_150k.txt
  target_words_vocabulary: combined_trg_vocab_150k.txt
  train_alignments:
    -  euro_corpus_en_es.gdfa
    -  custom_corpus_en_es.gdfa

I'm using `--model Transformer` 

Is this correct way to test this functionality ...

Thanks !

**Comment by guillaumekln** [comment]
Hi,

> Can I just give two eval files one for each train file ?

Weighted dataset is a training only feature and is independent from the evaluation files. You can only pass a single evaluation file for now.

> Do I have to combine the vocabulary sets from both train files for source_words_vocabulary and target_words_vocabulary ?

Most likely yes, otherwise you could get more OOV tokens for one dataset.

> For train_alignments do I have to generate one or two files ?

Training alignments should be aligned to the training files, so you should generate as many alignment files as there are training files.

**Comment by mohammedayub44** [comment]
I get this "invalid Argument Error:" ..not sure what is causing this - 

![image](https://user-images.githubusercontent.com/7122670/75281285-fe5ab300-57dc-11ea-9580-7d964f1e5d19.png)


My config is as below:
![image](https://user-images.githubusercontent.com/7122670/75281491-690bee80-57dd-11ea-9127-9d60c8368129.png)

I'm using your latest OpenNMT-tf 2.7.0 and Tensorflow 2.1.0

**Comment by guillaumekln** [comment]
This usually happens when there is an error in the alignment files, e.g. they are not aligned with the training file, or one line is longer than the training example. Can you check for that?

Actually, are you using the alignments during translation? If not, maybe it's just safer to remove them for now.

**Comment by mohammedayub44** [comment]
I have used the same files for my previous models and they worked fine. I will take a deeper look later. Any good way to troubleshoot these alignment files ? I might as well recreate them.

> Actually, are you using the alignments during translation? If not, maybe it's just safer to remove them for now.

I presently use them just to plot the attention matrix in the results as a verification. Not that critical. 
I have removed them for now and the training works. Will update in sometime.

**Comment by mohammedayub44** [comment]
Any reason why rouge metric is giving this error.. I mostly use BLEU but wanted to compare against rouge..also I see BLEU-detok is no more an option.

![image](https://user-images.githubusercontent.com/7122670/75412639-17ed1f00-58f1-11ea-85bb-7862518fafdc.png)

**Comment by guillaumekln** [comment]
> Any good way to troubleshoot these alignment files ?

You could write a script that compares the alignment file against the training file and detect mismatches (e.g. an ID that is larger than the sequence length).

We could also update the code to provide more context on errors. I also need to check if the dataset weighting is correctly applied to the alignment files.

> Any reason why rouge metric is giving this error.. I mostly use BLEU but wanted to compare against rouge..also I see BLEU-detok is no more an option.

BLEU-detok is the default in v2 and simply called BLEU. For ROUGE, I'm not exactly sure about this error but note that this is a metric for summarization tasks.

**Comment by mohammedayub44** [comment]
> We could also update the code to provide more context on errors. I also need to check if the dataset weighting is correctly applied to the alignment files.

Sounds good. That would be helpful.

> BLEU-detok is the default in v2 and simply called BLEU. For ROUGE, I'm not exactly sure about this error but note that this is a metric for summarization tasks.

Ah okay.

**Comment by mohammedayub44** [comment]
Thanks for pushing the extra logging changes. 
Also, looks like mixing in and out-of-domain data with weighted files drops BLEU scores drastically (My eval file has just 1000 sentences , 500 from each in and out of domain.)

![image](https://user-images.githubusercontent.com/7122670/75474685-6a682300-5965-11ea-90b8-7e6db8e842fc.png)

**Comment by guillaumekln** [comment]
Did you let this training run for longer? If yes, is there anything to report?

**Comment by mohammedayub44** [comment]
Yes. I removed the alignment files for now and ran the training for long time. I weighted 70:30 on base:custom on my input files. I used the merged vocabulary list. Because of the nature of my dataset it seemed to be overfitting very soon and BLUE scores dropping. 
Not sure of exact reason.  
I could try again with alignment files or different language pair.
